### DISCUSSION TITLE
VPS: ReferenceError: process is not defined

### DISCUSSION THREAD

**Discussion by iMrDJAi** [nextauthjs/next-auth#4566] [Help] [answered]
### Environment

```
  Binaries:
    Node: 16.15.0 - /usr/bin/node
    npm: 8.5.5 - /usr/bin/npm
  Browsers:
    Firefox: 99.0
  npmPackages:
    next-auth: ^4.3.4 => 4.3.4 
    react: ^17.0.2 => 17.0.2
    vite-plugin-ssr: ^0.4.0-beta.16 => 0.4.0-beta.16
```

### Reproduction URL

null

### Describe the issue

I'm using vite-plugin-ssr with NextAuth.js, I'm facing the error `ReferenceError: process is not defined` whenever I import anything from `next-auth/react` into my client side code (`index.page.tsx`), this prevents me from using the client API.

It looks like that NextAuth is directly using `process.env` from the client side: https://github.com/nextauthjs/next-auth/blob/main/packages/next-auth/src/react/index.tsx#L49. As we know, `process.env` is not available on browser, this is what causes this error.

Next.js does indeed expose a `process.env` object (https://nextjs.org/docs/basic-features/environment-variables). However, this is not the case for VPS and other frameworks that may work with NextAuth.

As a workaround, add this line to your HTML:

```html
<script>self.process = { env: {} }</script>
```

### How to reproduce

null

### Expected behavior

null

**Comment by balazsorban44** [comment]
`process` is polyfilled by Next.js, this should work. `vite-plugin-ssr` is not officially supported by NextAuth.js currently. We welcome PRs though. :+1:

**Reply by brillout** [reply]
It does work though :-).

Many use vite-plugin-ssr together with NextAuth.js.

FYI the vite-plugin-ssr docs for it: https://vite-plugin-ssr.com/NextAuth.js &mdash; in case you'd want to add a link to it in NextAuth.js's docs. (I can also do a PR if you want. Just checked: I'd add a guide, and an entry to `FAQ > Compatibility > Can I use NextAuth.js with Vite?`.)

**Comment by iMrDJAi** [comment]
@balazsorban44 It's unclear to me how Next.js is exposing environment variables to the client like `process.env.NEXTAUTH_URL` even though they are not prefixed by [NEXT_PUBLIC_](https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser). Am I missing something?

**Comment by Sajad-Sharhani** [comment]
install `dotenv` and add this to vite.config
```js
define: {
    "process.env": process.env,
  }
```

**Reply by brillout** [reply]
I'm surprised this works. Is that exactly what you are doing?
 
Also, be careful with such wildcard solutions. If you inadvertently have a `process.env.NEXTAUTH_SECRET` somewhere in a `**/*.page.js` dependency then you'll leak it. (E.g. if you inadvertently import a serve-side NextAuth.js module.)

**Comment by iMrDJAi** [comment]
@Sajad-Sharhani Defining the `process.env` object itself is not the issue. The issue is how to manage to pass `process.env.NEXTAUTH_URL_INTERNAL`, `process.env.NEXTAUTH_URL`, `process.env.VERCEL_URL` to the client side? (https://github.com/nextauthjs/next-auth/blob/main/packages/next-auth/src/react/index.tsx#L49)

What I don't understand is that these variables are not prefixed by `NEXT_PUBLIC_`. What's going on under the hood?

**Reply by brillout** [reply]
You can configure Vite to make these available on the client-side, see https://vite-plugin-ssr.com/NextAuth.js.

**Comment by brillout** [comment, marked as answer]
Check the official NextAuth.js vite-plugin-ssr docs at https://vite-plugin-ssr.com/NextAuth.js.

**Comment by iMrDJAi** [comment]
@brillout Man that's awesome! So exposing environment variables with custom prefixes to the client is possible in Vite!


Since you've updated the docs, consider mentioning that a fetch polyfill in node is needed in order to use `getSession()` and other related functions on server side (I recommend `cross-fetch`). Next.js provides a fetch polyfill by default.

```ts
import fetch, { Headers, Request, Response } from 'cross-fetch'

if (!globalThis.fetch) {
  globalThis.fetch = fetch
  globalThis.Headers = Headers
  globalThis.Request = Request
  globalThis.Response = Response
}
```

**Reply by brillout** [reply]
Docs updated https://vite-plugin-ssr.com/NextAuth.js#fetch-polyfill-for-getsession.

Do you know whether a polyfill is still needed for Node v18? That'd be great if not.

Also, just to be clear, the polyfill is not needed on the client-side right? Why not use `node-fetch` instead of `cross-fetch` then? Doesn't `cross-fetch` use `node-fetch` under the hood?

Also a thing we can do is to add an example to vite-plugin-ssr's `examples/` directory, which will be tested against upon each vite-plugin-ssr commit ensuring NextAuth.js support is here to stay. Let me know if you'd be up for a PR.

**Reply by iMrDJAi** [reply]
@brillout Since `node-fetch` is ESM only, you can't import it directly to your project. I tried but I couldn't get it working, I'm not sure if this issue is related to my configuration or not.

```
Error [ERR_REQUIRE_ESM]: require() of ES Module /home/djamel/project/node_modules/node-fetch/src/index.js from /home/djamel/project/server/index.ts not supported.
Instead change the require of index.js in /home/djamel/project/server/index.ts to a dynamic import() which is available in all CommonJS modules.
```

I'm using ESM though, I'm sure that the code is getting transpiled under the hood, maybe ts-node is the one to blame?

Node v18 doesn't require a pollyfill, but it's too early to rely on that.

I'm up to add a NextAuth.js example with custom credentials, custom register/login pages, and a way to pass cookies/session object to pages. Once it's done I'll do a PR!

**Reply by brillout** [reply]
> I'm using ESM though

Do you also have `package.json#type` set to `"module"`? That's probably why. And yes also make sure your tsconfig is set to spit out ESM.

Docs updated.

Looking forward to your PR.

**Reply by iMrDJAi** [reply]
@brillout I decided to stick to `cross-fetch` to avoid TypeScript issues since I'm using the DOM lib. node-fetch and native fetch types are not compatible.

**Reply by brillout** [reply]
👍 Let's use `cross-fetch` in the official example then.

**Reply by haftav** [reply]
Following up on this discussion, I believe if you stay on a v2 `node-fetch` version that should work as well. Looks like `node-fetch` v3 is [ESM only](https://github.com/node-fetch/node-fetch/blob/main/docs/v3-UPGRADE-GUIDE.md#converted-to-es-module).

**Reply by brillout** [reply]
@haftav Yes I can confirm that.

**Reply by iMrDJAi** [reply]
@brillout My example is done, please have a look: https://github.com/iMrDJAi/vps-nextauth-example

**Reply by brillout** [reply]
@iMrDJAi That's great! I've added your example to https://vite-plugin-ssr.com/NextAuth.js#examples.
